### PR TITLE
Implement From for more types on Cow

### PR DESCRIPTION
**Pull request by burtonageo**
This is basically https://github.com/rust-lang/rust/pull/48191, except that it should be implemented in a way that doesn't break third party crates.

### CODE REVIEW

**Comment by burtonageo** [comment]
`pathbuf_from_cow_path`, `cstring_from_cow_cstr` and `osstring_from_cow_osstr` all added

**Comment by KodrAus** [comment]
Thanks @burtonageo! This looks good to me.

@rfcbot fcp merge

These conversions are insta-stable, but I don't think they're contentious. We now have a pattern with our `From` conversions for `Cow` between `T` and `T::Owned`:

```
impl<'a, T> From<Cow<'a, T>> for T::Owned
impl<'a, T> From<&'a T> for Cow<'a, T> 
impl<'a, T> From<T::Owned> for Cow<'a, T> 
impl<'a, T> From<&'a T::Owned> for Cow<'a, T>
```

This brings our set of `From` conversions to:

```diff
  impl<'a, T> From<Cow<'a, [T]>> for Vec<T>
  impl<'a, T> From<&'a [T]> for Cow<'a, [T]> 
  impl<'a, T> From<Vec<T>> for Cow<'a, [T]> 
+ impl<'a, T> From<&'a Vec<T>> for Cow<'a, [T]>

  impl<'a> From<Cow<'a, str>> for String
  impl<'a> From<&'a str> for Cow<'a, str>
  impl<'a> From<String> for Cow<'a, str>
+ impl<'a> From<&'a String> for Cow<'a, str>

+ impl<'a> From<Cow<'a, Path>> for PathBuf
  impl<'a> From<&'a Path> for Cow<'a, Path>
  impl<'a> From<PathBuf> for Cow<'a, Path>
+ impl<'a> From<&'a PathBuf> for Cow<'a, Path>

+ impl<'a> From<Cow<'a, CStr>> for CString
+ impl<'a> From<&'a CStr> for Cow<'a, CStr>
+ impl<'a> From<CString> for Cow<'a, CStr>
+ impl<'a> From<&'a CString> for Cow<'a, CStr>

+ impl<'a> From<Cow<'a, OsStr>> for OsString
+ impl<'a> From<&'a OsStr> for Cow<'a, OsStr>
+ impl<'a> From<OsString> for Cow<'a, OsStr>
+ impl<'a> From<&'a OsString> for Cow<'a, OsStr>
```

**Comment by KodrAus** [comment]
@rfcbot fcp merge

**Comment by KodrAus** [comment]
One more time...

@rfcbot fcp merge

**Comment by KodrAus** [comment]
@rfcbot we are no longer friends.

I could just r+ this but figured I'd use rfcbot since these impls won't have an unstable period.

cc @rust-lang/libs

**Comment by SimonSapin** [comment]
@rfcbot fcp merge

File: src/liballoc/string.rs [hunk 1]
```diff
@@ -2239,6 +2239,14 @@ impl<'a> From<String> for Cow<'a, str> {
     }
 }
 
+#[stable(feature = "cow_from_string_ref", since = "1.27.0")]
```
Comment (kennytm) [inline src/liballoc/string.rs]: These should all become 1.28.0

**Comment by burtonageo** [comment]
@KodrAus Can this now enter final comment period?

**Comment by kennytm** [comment]
@burtonageo  

It needs one more vote from either @Kimundi, @aturon or @withoutboats.

**Comment by TimNN** [comment]
Ping, @Kimundi, @aturon or @withoutboats: This PR has an open FCP request that needs your input.

**Comment by oli-obk** [comment]
This PR did still cause breakage: https://github.com/rust-lang/rust/issues/51844

**Pull request merged**